amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerSubscriberTest.java

```diff
@@ -311,6 +309,80 @@ public void restartAfterRequestTimerExpiresTest() throws Exception {
 
     }
 
+    @Test
+    public void restartAfterRequestTimerExpiresAfterInitialSubscriptionFailsTest() throws Exception {
+
+        executorService = Executors.newFixedThreadPool(1, new ThreadFactoryBuilder()
+                .setNameFormat("test-" + testName.getMethodName() + "-%04d").setDaemon(true).build());
+
+        recordsPublisher = new RecordPublisherWithInitialFailureSubscription();
+        subscriber = new ShardConsumerSubscriber(recordsPublisher, executorService, bufferSize, shardConsumer, 0);
+        addUniqueItem(1);
+        addTerminalMarker(1);
+
+        CyclicBarrier barrier = new CyclicBarrier(2);
+
+        List<ProcessRecordsInput> received = new ArrayList<>();
+        doAnswer(a -> {
+            ProcessRecordsInput input = a.getArgumentAt(0, ProcessRecordsInput.class);
+            received.add(input);
+            if (input.records().stream().anyMatch(r -> StringUtils.startsWith(r.partitionKey(), TERMINAL_MARKER))) {
+                synchronized (processedNotifier) {
+                    processedNotifier.notifyAll();
+                }
+            }
+            return null;
+        }).when(shardConsumer).handleInput(any(ProcessRecordsInput.class), any(Subscription.class));
+
+        synchronized (processedNotifier) {
+            subscriber.startSubscriptions();
+            processedNotifier.wait(1000);
+        }
+
+        Stream.iterate(2, i -> i + 1).limit(97).forEach(this::addUniqueItem);
+
+        addTerminalMarker(2);
+
+        verify(shardConsumer, times(0)).handleInput(argThat(eqProcessRecordsInput(processRecordsInput)),
+                any(Subscription.class));
+
+        synchronized (processedNotifier) {
+            assertThat(subscriber.healthCheck(1), nullValue());
+            processedNotifier.wait(5000);
+        }
+
+        synchronized (processedNotifier) {
+            executorService.execute(() -> {
+                try {
+                    //
+                    // Notify the test as soon as we have started executing, then wait on the post add
+                    // subscriptionBarrier.
+                    //
+                    synchronized (processedNotifier) {
+                        processedNotifier.notifyAll();
+                    }
+                    barrier.await();
+                } catch (Exception e) {
+                    log.error("Exception while blocking thread", e);
+                }
+            });
+            //
+            // Wait for our blocking thread to control the thread in the executor.
+            //
+            processedNotifier.wait(5000);
+        }
+
+        barrier.await(500, TimeUnit.MILLISECONDS);
+
+        verify(shardConsumer, times(100)).handleInput(argThat(eqProcessRecordsInput(processRecordsInput)),
+                any(Subscription.class));
+
+        assertThat(received.size(), equalTo(recordsPublisher.responses.size()));
+        Stream.iterate(0, i -> i + 1).limit(received.size()).forEach(i -> assertThat(received.get(i),
```

Comment:
Could you add a comment for what this test is verifying? It's a little hard to follow

Reply:
Added comments and simplifying the test too.